Code/Editor/CryEditPy.cpp

```diff
@@ -150,23 +150,23 @@ namespace
 
         auto previousDocument = GetIEditor()->GetDocument();
         QString previousPathName = (previousDocument != nullptr) ? previousDocument->GetLevelPathName() : "";
-        auto newDocument = CCryEditApp::instance()->OpenDocumentFile(levelPath.toUtf8().data());
+        auto newDocument = CCryEditApp::instance()->OpenDocumentFile(levelPath.toUtf8().data(), addToMostRecent, reopenIfSameLevel);
 
         // the underlying document pointer doesn't change, so we can't check that; use the path name's instead
 
         bool result = true;
-        if (newDocument == nullptr || newDocument->IsLevelLoadFailed() || (newDocument->GetLevelPathName() == previousPathName))
+        if (newDocument == nullptr || newDocument->IsLevelLoadFailed() || (!reopenIfSameLevel && newDocument->GetLevelPathName() == previousPathName))
         {
             result = false;
         }
 
         return result;
     }
 
-    bool PyOpenLevelNoPrompt(const char* pLevelName)
+    bool PyOpenLevelNoPrompt(const char* pLevelName, bool addToMostRecent = false, bool reopenIfSameLevel = true)
```

Comment:
As a general rule, adding extra functionality by adding booleans to functions like this in not ideal. They tend to be confusing when they are called and normally they tend to mix different functionalities in one function. Normally a better approach would be to add another function to cover that case, it will include a different name, which will make it directly more intuitive to use, and if there is really a common code, that can be refactor to another function.

I suggest to rethink this PR considering other approaches than adding booleans to add functionality.

Reply:
This is a python API, it is not a problem for it because in python you will be able to do:

general.open_level("MyLevel", reopenIfSameLevel=False)

Reply:
This is C++ code, so potentially it will not only be called from python. And this pattern is not only done here, it goes all the way down to the `OpenDocumentFile` function from `CryEdit.h`, so it has spread already with wide impact to basic classes like `CCryDocManager`.

> One of the reasons of why not adding a new function is because it will require to modify every single test that uses open level

This means that this change is not trivial and a wide impact is going to be made. 

Maybe a middle ground could be to keep the booleans at higher levels (called from python) but split the different functionality in functions in lower levels.

Reply:
This will be exclusively called from python because:
- It is defined only in a Cpp file
- The file and the function are clearly only intended for being used from Python (CryEditPy.cpp and PyOpenLevel)

I can agree that this impacts CryEdit.h though. The biggest problem is that this is legacy code that god knows what would happen if I attempt to refactor the existing function. Adding the bool with default parameter is the less risky approach for the whole editor to make this change. In any case this would be replaced eventually so I don't think its too much of an issue to add this bool.
I can change this to be an enum so it is more clear when calling the function so you can't never do func(false, false)

This change has impact, but it will only impact python tests that reopen the same level. Reopening the same level will always be used with the intention of reloading the level content.

Reply:
Thanks for the explanation. Let's see also what other people think as well. You've got a point the risk of refactoring legacy code.

> I can change this to be an enum so it is more clear when calling the function so you can't never do func(false, false)

That would help for the CryEdit function.

Reply:
What do you think of adding a function `CCryDocManager::ReOpenDocumentFile`?

That way internally only need to do something like this
````
CloseDocumentFile(...);
OpenDocumentFile(...);
````
Then higher levels will call Open or Reopen version depending on the boolean.

Reply:
Despite this is old code, I wouldn't consider it Legacy. The parts of the Editor that will go away are old functionality UIs/plugins, I don't think anyone has to refactor/remove the core parts of the Editor in their radars.
@bosnichd ?

I like @moraaar suggestion, but would stick with Close/Open. If the Python side needs to do a reopen, they can just call Close/Open. This can also cleanup the "dont prompt" variable from `open_level`, that should only be in Close (that variable doesnt make sense in an Open if there isn't anything currently opened).

Reply:
> Despite this is old code, I wouldn't consider it Legacy. The parts of the Editor that will go away are old functionality UIs/plugins, I don't think anyone has to refactor/remove the core parts of the Editor in their radars.
> @bosnichd ?
> 
> I like @moraaar suggestion, but would stick with Close/Open. If the Python side needs to do a reopen, they can just call Close/Open. This can also cleanup the "dont prompt" variable from `open_level`, that should only be in Close (that variable doesnt make sense in an Open if there isn't anything currently opened).

I'm not aware of any plans to refactor or remove anything from Code/Editor, aside from what would be necessary in order to eventually remove Code/Legacy, although anything in Code/Editor that is clearly unused should be fair game for removal. Doesn't sound like either applies to this case, although it does seem tangentially related to the eventual removal/replacement of Code/Legacy/CrySystem/LevelSystem - when that happens it may be a good time to make whatever changes we deem desirable/necessary to these editor/python specific level open/close APIs?

Reply:
@amznestebanpapp CCryEditApp does not have a CloseDocument() functionality unfortunately from what I see in the code. There is no editor option for closing a level, as it is expecting to always have a level. Note that if you click cancel on startup and don't select a level it stays in this weird "no level" mode, but that is the only way to achieve it.
We could add CloseDocument in the future but I think this may get out of scope of the initial intention of this PR which is adding stability for batched tests.

Also, reopening the same level is a custom functionality that totally is very subjective if its even needed to be done.
If I want to discard my changes in my level and reopen it, I need to open a different level and then open the previous one.
That's why I believe at least, the python scripts should always open the level from scratch.

Reply:
I see, ok, then we can aim to something like that longer term and roll with this in the mean time